chain/jsonrpc/src/lib.rs

```diff
@@ -114,112 +118,116 @@ fn parse_hash(params: Option<Value>) -> Result<CryptoHash, RpcError> {
+    async fn send_tx_async(&self, params: Option<Value>) -> Result<Value, RpcError> {
+        let tx = parse_tx(params)?;
         let hash = (&tx.get_hash()).into();
         actix::spawn(
             self.client_addr
                 .send(NetworkClientMessages::Transaction(tx))
                 .map(|_| ())
                 .map_err(|_| ()),
         );
-        Box::new(future::ok(Value::String(hash)))
+        Ok(Value::String(hash))
     }
 
-    fn send_tx_commit(&self, params: Option<Value>) -> Box<Future<Item = Value, Error = RpcError>> {
-        let tx = ok_or_rpc_error!(parse_tx(params));
+    async fn send_tx_commit(&self, params: Option<Value>) -> Result<Value, RpcError> {
+        let tx = parse_tx(params)?;
         let tx_hash = tx.get_hash();
-        let view_client_addr = self.view_client_addr.clone();
         let RpcPollingConfig { polling_interval, polling_timeout, .. } = self.polling_config;
-        Box::new(
-            self.client_addr
-                .send(NetworkClientMessages::Transaction(tx))
-                .map_err(|err| RpcError::server_error(Some(err.to_string())))
-                .and_then(move |_result| {
-                    Interval::new_interval(polling_interval)
-                        .map(move |_| view_client_addr.send(TxStatus { tx_hash }).wait())
-                        .filter(|tx_status| match tx_status {
-                            Ok(Ok(tx_status)) => match tx_status.status {
-                                FinalTransactionStatus::Started
-                                | FinalTransactionStatus::Unknown => false,
-                                _ => true,
-                            },
-                            _ => false,
-                        })
-                        .map_err(|err| RpcError::server_error(Some(err.to_string())))
-                        .take(1)
-                        .fold(Value::Null, |_, tx_status| jsonify(tx_status))
-                })
-                .timeout(polling_timeout)
-                .map_err(|_| {
-                    RpcError::server_error(Some("send_tx_commit has timed out.".to_owned()))
-                }),
-        )
+        self.client_addr
+            .send(NetworkClientMessages::Transaction(tx))
+            .map_err(|err| RpcError::server_error(Some(err.to_string())))
+            .compat()
+            .await?;
+        let final_tx_async = async {
+            loop {
+                let final_tx = self.view_client_addr.send(TxStatus { tx_hash }).compat().await;
+                if let Ok(Ok(ref tx)) = final_tx {
+                    match tx.status {
+                        FinalTransactionStatus::Started | FinalTransactionStatus::Unknown => {}
+                        _ => {
+                            break jsonify(final_tx);
+                        }
+                    }
+                }
+                let _ = delay(polling_interval).await;
+            }
+        };
+        select! {
```

Comment:
Could you explain what this code and `select!` macro itself is doing and why we cannot return `final_tx_async.await` directly?

Reply:
@nearmax It is a timeout implementation. Given you asked, it is an indicator that I need to extract it into a dedicated function.

Reply:
@nearmax I have extracted it into a separate helper function.

Do you know the best place where these helpers can go?

Reply:
I think in this case we can create async_utils crate and put it in there.